probe/plugins/registry.go

```diff
@@ -180,23 +210,140 @@ func (r *Registry) Name() string { return "plugins" }
+		&rpt.ReplicaSet,
+		&rpt.Host,
+		&rpt.Overlay,
+	}
+}
+
+func (r *Registry) rmPluginHandlers(pluginID string) {
+	for _, controlID := range r.controlsByPlugin[pluginID] {
+		fakeID := fakeControlID(pluginID, controlID)
+		controls.RmLocked(fakeID)
+	}
+	delete(r.controlsByPlugin, pluginID)
+}
+
+func (r *Registry) updateAndRegisterControlsInTopology(pluginID string, topology *report.Topology) {
+	var pluginControls []string
+	newControls := report.Controls{}
+	for controlID, control := range topology.Controls {
+		fakeID := fakeControlID(pluginID, controlID)
+		log.Debugf("plugins: replacing control %s with %s", controlID, fakeID)
+		control.ID = fakeID
+		newControls.AddControl(control)
+		pluginControls = append(pluginControls, controlID)
+		controls.RegisterLocked(fakeID, r.pluginControlHandler)
+	}
+	newNodes := report.Nodes{}
+	for name, node := range topology.Nodes {
+		log.Debugf("plugins: checking node controls in node %s of %s", name, topology.Label)
+		newNode := node.WithID(name)
+		newLatestControls := report.MakeEmptyLatestNodeControls()
+		node.LatestControls.ForEach(func(controlID string, ts time.Time, data report.NodeControlData) {
+			log.Debugf("plugins: got node control %s", controlID)
+			newControlID := ""
+			if _, found := topology.Controls[controlID]; !found {
+				log.Debugf("plugins: node control %s does not exist in topology controls", controlID)
+				newControlID = controlID
+			} else {
+				newControlID = fakeControlID(pluginID, controlID)
+				log.Debugf("plugins: will replace node control %s with %s", controlID, newControlID)
+			}
+			newLatestControls = newLatestControls.Set(newControlID, ts, data)
+		})
+		newNode.LatestControls = newLatestControls
+		newNodes[newNode.ID] = newNode
+	}
+	topology.Controls = newControls
+	topology.Nodes = newNodes
+	r.controlsByPlugin[pluginID].Add(pluginControls...)
+}
+
+// PluginResponse is an extension of xfer.Response that allows plugins
+// to send the shortcut reports
+type PluginResponse struct {
+	xfer.Response
+	ShortcutReport *report.Report `json:"shortcutReport,omitempty"`
+}
+
+func (r *Registry) pluginControlHandler(req xfer.Request) xfer.Response {
+	pluginID, controlID := realPluginAndControlID(req.Control)
+	req.Control = controlID
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	if plugin, found := r.pluginsByID[pluginID]; found {
+		response := plugin.Control(req)
+		if response.ShortcutReport != nil {
+			r.updateAndRegisterControlsInReport(response.ShortcutReport)
+			response.ShortcutReport.Shortcut = true
+			r.publisher.Publish(*response.ShortcutReport)
+		}
+		return response.Response
+	}
+	return xfer.ResponseErrorf("plugin %s not found", pluginID)
+}
+
+func realPluginAndControlID(fakeID string) (string, string) {
+	parts := strings.SplitN(fakeID, "~", 2)
+	if len(parts) != 2 {
+		return "", fakeID
+	}
+	return parts[0], parts[1]
+}
+
+func fakeControlID(pluginID, controlID string) string {
+	return fmt.Sprintf("%s~%s", pluginID, controlID)
```

Comment:
No, there is only a check for plugin IDs not including `~`. Control IDs can contain `~` - we split the fake ID only against the first tilde.